util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,461 @@
+			tsr.Report(data)
+		case <-tsr.quit:
+			return
+		}
+	}
+}
+
+func (tsr *RemoteTopSQLReporter) Report(data reportData) {
+	batch := tsr.prepareReportData(data)
+	addr := variable.TopSQLVariable.AgentAddress.Load()
+	ctx, cancel := context.WithTimeout(context.TODO(), reportTimeout)
+	err := tsr.client.Send(ctx, addr, batch)
+	if err != nil {
+		logutil.BgLogger().Warn("[top-sql] client failed to send data", zap.Error(err))
+	}
+	cancel()
+}
+
+// ReportGRPCClient reports data to grpc servers.
+type ReportGRPCClient struct {
+	addr   string
+	conn   *grpc.ClientConn
+	client tipb.TopSQLAgentClient
+}
+
+// NewReportGRPCClient returns a new ReportGRPCClient
+func NewReportGRPCClient() *ReportGRPCClient {
+	return &ReportGRPCClient{}
+}
+
+// Send implements the ReportClient interface.
+func (r *ReportGRPCClient) Send(ctx context.Context, addr string, batch []*tipb.CollectCPUTimeRequest) error {
+	if addr == "" {
+		return nil
+	}
+	err := r.initialize(addr)
+	if err != nil {
+		return err
+	}
+	// TODO: test timeout behavior
+	stream, err := r.client.CollectCPUTime(ctx)
+	if err != nil {
+		return r.resetClientWhenSendError(err)
+	}
+	if err := r.sendBatch(stream, batch); err != nil {
+		return r.resetClientWhenSendError(err)
+	}
+	return nil
+}
+
+// sendBatch sends a batch of TopSQL records streamingly.
+// TODO: benchmark test with large amount of data (e.g. 5000), tune with grpc.WithWriteBufferSize()
+func (tsr *ReportGRPCClient) sendBatch(stream tipb.TopSQLAgent_CollectCPUTimeClient, batch []*tipb.CollectCPUTimeRequest) error {
+	for _, req := range batch {
+		if err := stream.Send(req); err != nil {
+			return err
+		}
+	}
+	// response is Empty, drop it for now
+	_, err := stream.CloseAndRecv()
+	return err
+}
+
+func (r *ReportGRPCClient) initialize(addr string) (err error) {
+	if r.addr == addr {
+		return nil
+	}
+	r.conn, r.client, err = r.newAgentClient(addr)
+	if err != nil {
+		return err
+	}
+	r.addr = addr
+	return nil
+}
+
+func (r *ReportGRPCClient) newAgentClient(addr string) (*grpc.ClientConn, tipb.TopSQLAgentClient, error) {
+	dialCtx, cancel := context.WithTimeout(context.Background(), dialTimeout)
+	defer cancel()
+	conn, err := grpc.DialContext(
+		dialCtx,
+		addr,
+		grpc.WithInsecure(),
+		grpc.WithInitialWindowSize(grpcInitialWindowSize),
+		grpc.WithInitialConnWindowSize(grpcInitialConnWindowSize),
+		grpc.WithDefaultCallOptions(
+			grpc.MaxCallRecvMsgSize(math.MaxInt64),
+		),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoff.Config{
+				BaseDelay:  100 * time.Millisecond, // Default was 1s.
+				Multiplier: 1.6,                    // Default
+				Jitter:     0.2,                    // Default
+				MaxDelay:   3 * time.Second,        // Default was 120s.
+			},
+		}),
+	)
+	if err != nil {
+		return nil, nil, err
+	}
+	client := tipb.NewTopSQLAgentClient(conn)
+	return conn, client, nil
+}
+
+func (r *ReportGRPCClient) resetClientWhenSendError(err error) error {
```

Comment:
Why do we need to reset it instead of letting gRPC to handle it?

Reply:
The error was return from grpc, if gRPC can handle it by backoff, it won't return the error.